### ISSUE TITLE
Reworking the `parent` property on AST Nodes

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#1417]
Right now the `parent` property is optional on all AST nodes.
However in reality, it's always defined on all AST nodes, except for `Program`, where it is missing.

This leads to a bunch of unnecessary ifs/optional chains/"trust me it's right" assertions, all of which "harm" the type safety, or are impossible to get coverage on (leading to istanbul ignore statements):

```ts
// this shouldn't happen, checking just in case
/* istanbul ignore if */ if (node.parent) {
  // ...
}

// selector asserts this is correct
const parent = node.parent as TSESTree.VariableDeclaration;

if (node.parent?.type === AST_NODE_TYPES.VariableDeclaration) {
  // ...
}
```

There is an issue with this, however.
`typescript-estree` does not assign the parent property; this is done by ESLint as it traverses the AST at lint time.

So there are two solutions I can see:
1) change `typescript-estree` so that it sets the `parent` appropriately.
2) create a separate set of types for use within plugins, which correctly type the `parent` property.

---

As an aside, doing this would unlock a really awesome benefit wherein we could make all the types generic, allowing you to specify the expected parent type ahead of time:
```ts
interface BaseNode<TParent extends Node = Node> {
  // ...
  parent: TParent;
}

interface VariableDeclarator<TParent extends Node = Node> extends BaseNode<TParent> {}

interface Program extends BaseNode {
  parent?: never;
}
```

This would be great for things like return values of functions which explicitly check parent types, and for eslint selectors that explicitly state the tree structure (`VariableDeclaration > VariableDeclarator`).

**Comment by armano2** [comment]
Note: Setting up parent within ts-estree can be misleading for projects that do not use eslint.

what do you think about using type augmentation inside eslint-utils/plugin that will extend nodes instead?

```ts
import { Node } from '@typescript-eslint/typescript-estree/dist/ts-estree/ts-estree';

declare module '@typescript-eslint/typescript-estree/dist/ts-estree/ts-estree' {
  export interface BaseNode {
    parent: Node;
  }

  export interface Program {
    parent: never;
  }
}
```

https://github.com/typescript-eslint/typescript-eslint/pull/1416/files#diff-ea705e9b27753be392a33703d63866ce

**Comment by bradzacher** [comment]
> Note: Setting up parent within ts-estree can be misleading for projects that do not use eslint.

Hence point (1) - changing `typescript-estree` so it does.

> what do you think about using type augmentation inside eslint-utils/plugin that will extend nodes instead?

Potentially, but the problem with that is that it doesn't help users of `experimental-utils`.
If we can make the declaration merging work for the types within the `experimental-utils` package, then we don't need to worry about modifying the types in `typescript-estree`.

**Comment by armano2** [comment]
i like idea of using templates for parent, it will be awesome if there will be way to do so from within `experimental-utils`, but i'm unsure if you can do augmentation that adds template to interfaces.

(i did my PR just as POC in 5min)

**Comment by G-Rath** [comment]
fwiw it looks like type argumentation no longer works because `@typescript-eslint/types` is itself using that to add the `parent` property - so the result is:

```
All declarations of 'parent' must have identical modifiers.(2687)
Subsequent property declarations must have the same type.  Property 'parent' must be of type 'Node | undefined', but here has type 'Node'.(2717)
```

[playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQMoFFXKgUx3AXzgDMoIQ4ByAARgE8wcBnAYymDBgFpmAbYAHYwA9AFcYwXk0oBuAFByAJjha8AhrjggIi0b3w16jVu048m-IcKPNhi4ExEBzHAJxQ1MHIuFrHXJkYWSkQ5ODgcAA9IWDhBLyhiNRZ8ACE-HAA5HXwEMPC4MA1XGAAuFAwsXBwAOmzleQKtOgAFMkZYOnLHdgEnRsI5AgUWCAFHEgE4AF44AAoBHPK0TGw8OpyAShmAPlCm0fGIfRreCCcFnJqQVvb3ek35AiA).

I can't find anyway around this, though would be good if I'm wrong 😅

**Comment by bradzacher** [comment]
Yeah!

When I sit out the ast-spec package I intentionally did not add the parent property because it's not part of the spec!

It's added here
https://github.com/typescript-eslint/typescript-eslint/blob/c4310b1aac35c7d31b826f0602eca6a5900a09ee/packages/types/src/ts-estree.ts

Thanks for reminding me though - we can make the breaking change in the next major!

**Comment by JoshuaKGoldberg** [comment]
#5252 is merged into the `v6` branch 🥳. Closing this as it'll be resolved in v6.
